File: clients/cpp/tests/test_client.cpp
#include <gtest/gtest.h>
#include <dbforge/dbforge.hpp>

class DBForgeClientTest : public ::testing::Test {
protected:
    void SetUp() override {
        // These tests require a running DB-Forge instance
        // Skip tests if server is not available
    }
    
    void TearDown() override {
        // Cleanup any test databases
    }
};

TEST_F(DBForgeClientTest, ClientConstruction) {
    EXPECT_NO_THROW({
        dbforge::Client client("http://test.localhost", "test-key", 30);
        EXPECT_EQ(client.base_url(), "http://test.localhost");
        EXPECT_EQ(client.api_key(), "test-key");
    });
}

TEST_F(DBForgeClientTest, ColumnConstruction) {
    dbforge::Column col1("id", "INTEGER");
    EXPECT_EQ(col1.name, "id");
    EXPECT_EQ(col1.type, "INTEGER");
    EXPECT_FALSE(col1.primary_key);
    EXPECT_FALSE(col1.not_null);
    
    dbforge::Column col2("username", "TEXT", true, true, "default_user");
    EXPECT_EQ(col2.name, "username");
    EXPECT_EQ(col2.type, "TEXT");
    EXPECT_TRUE(col2.primary_key);
    EXPECT_TRUE(col2.not_null);
    EXPECT_EQ(col2.default_value, "default_user");
}

TEST_F(DBForgeClientTest, ExceptionHierarchy) {
    try {
        throw dbforge::DatabaseNotFound("Test message", 404, "NOT_FOUND");
    } catch (const dbforge::DatabaseNotFound& e) {
        EXPECT_EQ(e.status_code(), 404);
        EXPECT_EQ(e.error_code(), "NOT_FOUND");
        EXPECT_STREQ(e.what(), "Test message");
    } catch (...) {
        FAIL() << "Wrong exception type caught";
    }
    
    try {
        throw dbforge::InvalidRequest("Bad request");
    } catch (const dbforge::Exception& e) {
        EXPECT_EQ(e.status_code(), 400);
        EXPECT_EQ(e.error_code(), "BAD_REQUEST");
    } catch (...) {
        FAIL() << "Exception not caught as base type";
    }
}

// Integration tests (require running server)
class DBForgeIntegrationTest : public ::testing::Test {
protected:
    void SetUp() override {
        try {
            client = std::make_unique<dbforge::Client>("http://db.localhost");
            // Try to connect to server
            client->health_check();
            server_available = true;
        } catch (...) {
            server_available = false;
            GTEST_SKIP() << "DB-Forge server not available, skipping integration tests";
        }
    }
    
    void TearDown() override {
        if (server_available && client) {
            // Cleanup test database if it exists
            try {
                client->prune_database("test_db");
            } catch (...) {
                // Ignore cleanup errors
            }
        }
    }
    
    std::unique_ptr<dbforge::Client> client;
    bool server_available = false;
};

TEST_F(DBForgeIntegrationTest, DatabaseLifecycle) {
    if (!server_available) return;
    
    // Spawn database
    auto spawn_result = client->spawn_database("test_db");
    EXPECT_FALSE(spawn_result.message.empty());
    EXPECT_EQ(spawn_result.database_name, "test_db");
    
    // List databases
    auto databases = client->list_databases();
    bool found = false;
    for (const auto& db : databases) {
        if (db.name == "test_db") {
            found = true;
            break;
        }
    }
    EXPECT_TRUE(found);
    
    // Get database instance
    auto db = client->get_database("test_db");
    EXPECT_EQ(db.name(), "test_db");
    
    // Prune database
    auto prune_result = client->prune_database("test_db");
    EXPECT_FALSE(prune_result.message.empty());
}

TEST_F(DBForgeIntegrationTest, TableOperations) {
    if (!server_available) return;
    
    client->spawn_database("test_db");
    auto db = client->get_database("test_db");
    
    // Create table
    std::vector<dbforge::Column> columns = {
        {"id", "INTEGER", true, false},
        {"name", "TEXT", false, true}
    };
    
    auto create_result = db.create_table("test_table", columns);
    EXPECT_FALSE(create_result.message.empty());
    
    // List tables
    auto tables = db.list_tables();
    EXPECT_EQ(tables.size(), 1);
    EXPECT_EQ(tables[0], "test_table");
    
    // Get schema
    auto schema = db.get_table_schema("test_table");
    EXPECT_EQ(schema.size(), 2);
    
    // Insert data
    std::vector<dbforge::Row> rows = {
        {{"name", "Alice"}},
        {{"name", "Bob"}}
    };
    
    auto insert_result = db.insert_rows("test_table", rows);
    EXPECT_EQ(insert_result.rows_affected, 2);
    
    // Select data
    auto selected = db.select_rows("test_table");
    EXPECT_EQ(selected.size(), 2);
    
    // Query with filter
    auto filtered = db.select_rows("test_table", {{"name", "Alice"}});
    EXPECT_EQ(filtered.size(), 1);
    EXPECT_EQ(filtered[0].at("name"), "Alice");
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: clients/python/tests/test_client.py
"""Tests for DB-Forge Python client."""

import pytest
import json
from unittest.mock import Mock, patch
from dbforge_client import DBForgeClient, DBForgeError, DatabaseNotFound


class TestDBForgeClient:
    """Test cases for DBForgeClient."""
    
    def setup_method(self):
        """Set up test fixtures."""
        self.client = DBForgeClient(
            base_url="http://test.localhost",
            api_key="test-key"
        )
    
    @patch('requests.Session.request')
    def test_spawn_database_success(self, mock_request):
        """Test successful database spawn."""
        mock_response = Mock()
        mock_response.ok = True
        mock_response.json.return_value = {
            "message": "Database instance spawned successfully.",
            "db_name": "test-db",
            "container_id": "abc123"
        }
        mock_request.return_value = mock_response
        
        result = self.client.spawn_database("test-db")
        
        assert result["db_name"] == "test-db"
        assert result["container_id"] == "abc123"
        mock_request.assert_called_once()
    
    @patch('requests.Session.request')
    def test_spawn_database_error(self, mock_request):
        """Test database spawn with error."""
        mock_response = Mock()
        mock_response.ok = False
        mock_response.status_code = 400
        mock_response.json.return_value = {
            "error": {
                "code": "BAD_REQUEST",
                "message": "Invalid database name.",
                "status": 400
            }
        }
        mock_request.return_value = mock_response
        
        with pytest.raises(DBForgeError):
            self.client.spawn_database("invalid-db")
    
    @patch('requests.Session.request')
    def test_list_databases(self, mock_request):
        """Test listing databases."""
        mock_response = Mock()
        mock_response.ok = True
        mock_response.json.return_value = [
            {
                "name": "db1",
                "container_id": "abc123",
                "status": "running"
            },
            {
                "name": "db2", 
                "container_id": "def456",
                "status": "running"
            }
        ]
        mock_request.return_value = mock_response
        
        result = self.client.list_databases()
        
        assert len(result) == 2
        assert result[0]["name"] == "db1"
        assert result[1]["name"] == "db2"
    
    @patch('requests.Session.request')
    def test_database_not_found(self, mock_request):
        """Test database not found error."""
        mock_response = Mock()
        mock_response.ok = False
        mock_response.status_code = 404
        mock_response.json.return_value = {
            "error": {
                "code": "NOT_FOUND",
                "message": "Database instance not found.",
                "status": 404
            }
        }
        mock_request.return_value = mock_response
        
        with pytest.raises(DatabaseNotFound):
            self.client.prune_database("nonexistent-db")
    
    def test_get_database(self):
        """Test getting database instance."""
        db = self.client.get_database("test-db")
        
        assert db.name == "test-db"
        assert db.client == self.client


class TestDBForgeDatabase:
    """Test cases for DBForgeDatabase."""
    
    def setup_method(self):
        """Set up test fixtures."""
        self.client = DBForgeClient(base_url="http://test.localhost")
        self.db = self.client.get_database("test-db")
    
    @patch('requests.Session.request')
    def test_create_table(self, mock_request):
        """Test table creation."""
        mock_response = Mock()
        mock_response.ok = True
        mock_response.json.return_value = {
            "message": "Table 'users' created successfully."
        }
        mock_request.return_value = mock_response
        
        columns = [
            {"name": "id", "type": "INTEGER", "primary_key": True},
            {"name": "username", "type": "TEXT", "not_null": True}
        ]
        
        result = self.db.create_table("users", columns)
        
        assert "created successfully" in result["message"]
        
        # Verify the request was made correctly
        call_args = mock_request.call_args
        assert call_args[1]["method"] == "POST"
        assert "/api/db/test-db/tables" in call_args[1]["url"]
    
    @patch('requests.Session.request')
    def test_insert_rows(self, mock_request):
        """Test row insertion."""
        mock_response = Mock()
        mock_response.ok = True
        mock_response.json.return_value = {
            "message": "Rows inserted successfully.",
            "rows_affected": 2
        }
        mock_request.return_value = mock_response
        
        rows = [
            {"username": "alice", "email": "alice@example.com"},
            {"username": "bob", "email": "bob@example.com"}
        ]
        
        result = self.db.insert_rows("users", rows)
        
        assert result["rows_affected"] == 2
    
    @patch('requests.Session.request')
    def test_select_rows(self, mock_request):
        """Test row selection."""
        mock_response = Mock()
        mock_response.ok = True
        mock_response.json.return_value = {
            "data": [
                {"id": 1, "username": "alice", "email": "alice@example.com"}
            ],
            "rows_affected": 1
        }
        mock_request.return_value = mock_response
        
        result = self.db.select_rows("users", {"username": "alice"})
        
        assert len(result) == 1
        assert result[0]["username"] == "alice"
    
    @patch('requests.Session.request')
    def test_execute_query(self, mock_request):
        """Test raw query execution."""
        mock_response = Mock()
        mock_response.ok = True
        mock_response.json.return_value = {
            "data": [{"count": 5}],
            "rows_affected": 1
        }
        mock_request.return_value = mock_response
        
        result = self.db.execute_query("SELECT COUNT(*) as count FROM users")
        
        assert result["data"][0]["count"] == 5


if __name__ == "__main__":
    pytest.main([__file__])